plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -238,10 +250,33 @@ public ConnectorPageSource createPageSource(
         Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(partitionData, partitionSpec);
 
         ImmutableList.Builder<IcebergColumnHandle> requiredColumnsBuilder = ImmutableList.builder();
+        Set<Integer> requiredColumnIds = new HashSet<>();
         requiredColumnsBuilder.addAll(icebergColumns);
+        icebergColumns.stream().map(IcebergColumnHandle::getId).forEach(requiredColumnIds::add);
         deleteFilterRequiredSchema.stream()
                 .filter(column -> !icebergColumns.contains(column))
-                .forEach(requiredColumnsBuilder::add);
+                .forEach(column -> {
+                    requiredColumnsBuilder.add(column);
+                    requiredColumnIds.add(column.getId());
+                });
+        icebergColumns.stream().filter(IcebergColumnHandle::isUpdateRowIdColumn).findFirst().ifPresent(updateRowIdColumn -> {
+            updateRowIdColumn.getColumnIdentity().getChildren().forEach(requiredColumnIdentity -> {
+                // The $row_id column is not read directly, but its children are
+                // the row position column, and some number of data columns needed for the update.
+                if (!requiredColumnIds.contains(requiredColumnIdentity.getId())) {
+                    Type columnType;
+                    if (requiredColumnIdentity.getId() == ROW_POSITION.fieldId()) {
+                        columnType = BIGINT;
+                    }
+                    else {
+                        columnType = toTrinoType(tableSchema.findType(requiredColumnIdentity.getId()), typeManager);
+                    }
+
+                    requiredColumnsBuilder.add(new IcebergColumnHandle(requiredColumnIdentity, columnType, ImmutableList.of(), columnType, Optional.empty()));
```

Comment:
Comment is missing here. Can it hurt?

We had a bug recently because comment was missing in some code path, where `IcebergColumnHandle` were compared with equality.

Should we do equivalent of `tableSchema.columns().stream().map(column -> getColumnHandle(column, typeManager))` for constructing the `IcebergColumnHandle`?

Reply:
> Can it hurt?

I'll include the comment, just to be safe